Add meta registration for foreach_maximum_.List

janeyx99 opened the pull request:
Will fix issues compiling for when amsgrad is True for Adam(W), see related failures in https://github.com/pytorch/benchmark/actions/runs/5628705163/job/15252867793

Also did some refactoring where common registrations could be deduplicated.

Test plan:
python test/inductor/test_compiled_optimizers.py -k test_adam

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105864



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

janeyx99 committed:
Add meta registration for foreach_maximum_.List

[ghstack-poisoned]

janeyx99 committed:
Update on "Add meta registration for foreach_maximum_.List"

Will fix issues compiling for when amsgrad is True for Adam(W), see related failures in https://github.com/pytorch/benchmark/actions/runs/5628705163/job/15252867793




[ghstack-poisoned]

crcrpar commented on torch/_meta_registrations.py:
```diff
@@ -2590,43 +2590,31 @@ def _check_foreach_binop_tensor_lists(self, other):
+        aten._foreach_sub.List,
+        aten._foreach_mul.List,
+        aten._foreach_div.List,
+        aten._foreach_maximum.List,
```
admitting I'm not familiar with meta registration at all but wouldn't `_foreach_minimum` be worth adding as well?

crcrpar commented on torch/_meta_registrations.py:
```diff
@@ -2590,43 +2590,31 @@ def _check_foreach_binop_tensor_lists(self, other):
+        aten._foreach_sub_.List,
+        aten._foreach_mul_.List,
+        aten._foreach_div_.List,
+        aten._foreach_maximum_.List,
```
ditto

janeyx99 replied:
Oki! Thanks :D

albanD commented on torch/_meta_registrations.py:
```diff
@@ -2590,43 +2590,31 @@ def _check_foreach_binop_tensor_lists(self, other):
-def meta__foreach_binop_list(self, other):
+def meta__foreach_binop__list(self, other):
     _check_foreach_binop_tensor_lists(self, other)
-    return [torch.empty_like(s) for s in self]
```
Is no return ok here?
I know it matches the actual impl but not sure compile will be happy with that haha

janeyx99 replied:
No return is what's expected from inplace foreach ops so this actually can't return stuff :/

albanD replied:
Ok ok

albanD reviewed (approved):
SGTM if CI is happy and minimum added.
Also lerp, etc?

janeyx99 commented:
lerp is weird and has its own registrations. I'll also add some tests.

janeyx99 committed:
Update on "Add meta registration for foreach_maximum_.List"

Will fix issues compiling for when amsgrad is True for Adam(W), see related failures in https://github.com/pytorch/benchmark/actions/runs/5628705163/job/15252867793

Also did some refactoring where common registrations could be deduplicated.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot merge
